compiler/rustc_mir_build/src/thir/cx/expr.rs

```diff
@@ -158,6 +159,100 @@ impl<'tcx> Cx<'tcx> {
+            // Special cased so that we can type check that the element
+            // type of the source matches the pointed to type of the
+            // destination.
+            ExprKind::Pointer {
+                source: self.mirror_expr(source),
+                cast: PointerCast::ArrayToPointer,
+            }
+        } else {
+            // check whether this is casting an enum variant discriminant
+            // to prevent cycles, we refer to the discriminant initializer
+            // which is always an integer and thus doesn't need to know the
+            // enum's layout (or its tag type) to compute it during const eval
+            // Example:
+            // enum Foo {
+            //     A,
+            //     B = A as isize + 4,
+            // }
+            // The correct solution would be to add symbolic computations to miri,
+            // so we wouldn't have to compute and store the actual value
+
+            let hir::ExprKind::Path(ref qpath) = source.kind else {
+                return ExprKind::Cast { source: self.mirror_expr(source)};
+            };
+
+            let res = self.typeck_results().qpath_res(qpath, source.hir_id);
+            let (discr_did, discr_offset, discr_ty, substs) = {
+                let ty = self.typeck_results().node_type(source.hir_id);
+                let ty::Adt(adt_def, substs) = ty.kind() else {
+                    return ExprKind::Cast { source: self.mirror_expr(source)};
+                };
+                let Res::Def(
+                            DefKind::Ctor(CtorOf::Variant, CtorKind::Const),
+                            variant_ctor_id,
+                        ) = res else {
+                            return ExprKind::Cast { source: self.mirror_expr(source)};
+                        };
+
+                let idx = adt_def.variant_index_with_ctor_id(variant_ctor_id);
+                let (d, o) = adt_def.discriminant_def_for_variant(idx);
+                use rustc_middle::ty::util::IntTypeExt;
+                let ty = adt_def.repr().discr_type();
+                let ty = ty.to_ty(tcx);
+                (d, o, ty, substs)
+            };
+
+            let param_env_ty = self.param_env.and(discr_ty);
+            let size = tcx
+                .layout_of(param_env_ty)
+                .unwrap_or_else(|e| {
+                    panic!("could not compute layout for {:?}: {:?}", param_env_ty, e)
+                })
+                .size;
+
+            let lit = ScalarInt::try_from_uint(discr_offset as u128, size).unwrap();
+            let kind = ExprKind::NonHirLiteral { lit, user_ty: None };
+            let offset = self.thir.exprs.push(Expr { temp_lifetime, ty: discr_ty, span, kind });
+
+            let source = match discr_did {
+                Some(did) => {
+                    // in case we are offsetting from a computed discriminant
+                    // and not the beginning of discriminants (which is always `0`)
+
```

Comment:
```suggestion
                // in case we are offsetting from a computed discriminant
                // and not the beginning of discriminants (which is always `0`)
                Some(did) => {
```